dv/cosim/spike_cosim.cc

```diff
@@ -71,40 +71,56 @@ SpikeCosim::SpikeCosim(const std::string &isa_string, uint32_t start_pc,
 char *SpikeCosim::addr_to_mem(reg_t addr) { return nullptr; }
 
 bool SpikeCosim::mmio_load(reg_t addr, size_t len, uint8_t *bytes) {
+  // Load from the ISS memory model, while checking if the access is
+  // consistent with the observed DUT memory acceses.
+  // Return false if any errors occur, or the checking fails.
+  // Within Spike (mmu_t::load_slow_path), this causes a load_access_fault.
+
+  // Spike does not differentiate between iside and dside memory accesses
+  // (as ibex does) so we make an educated guess which is which, and
+  // then after fetching the data, check against the recorded DUT access
+  // that should be queued as pending.
+  //
+  // Spike may attempt to access up to 8-bytes from the PC when fetching, so
+  // assume as a dside access when it falls outside that range.
+  // N.B. Heuristic
+  uint32_t pc = processor->get_state()->pc;
+  bool is_probably_dmem_access = (addr < pc || addr >= (pc + 8));
+  // Check we don't have a heuristic mismatch (this may be a false-positive but
+  // worth failing explicity if it is seen)
+  assert(!(is_probably_dmem_access && pending_iside_error));
```

Comment:
This also needs to check if the pending iside error address matches the incoming address (`pending_iside_err_addr == aligned_addr` check you can see below).

Maybe add a new `bool` that combines the two as it'll be used in a couple of places?

Reply:
Ah yes good point. Will do!